docs/flip/1777-default-dtype.md

```diff
@@ -0,0 +1,102 @@
+# FLIP: Default dtypes
+
+
+- Start Date: 2022-01-11
+- FLIP PR: [#1776](https://github.com/google/flax/pull/1776)
+- FLIP Issue: [#1777](https://github.com/google/flax/issues/1777)
+
+
+## Summary
+
+This FLIP proposes to replace the default dtype which is currently hard-coded to float32, and instead use the JAX type promotion results to derive a default dtype from the input and parameters of a layer.
+
+
+## Motivation
+
+Currently, linen Modules always produce float32 outputs regardless of input and parameter dtypes. Half-precision types like float16 and bfloat16 are supported by explicitly passing the half-precision type to each layer. The way this is currently implemented is that each layer has a dtype argument with float32 as the default value. The layer guarantees that this dtype will be the return type of the result returned by `__call__`.
+
+The current behavior is problematic and results in silent bugs especially for dtypes that do not fit inside float32 (complex, float64). Also the linen dtype behavior is significantly different from how NumPy and by extension JAX handle dtypes.
+
+
+### Dtypes in JAX
+
+JAX uses a NumPy inspired [dtype promotion](https://github.com/google/jax/blob/main/jax/_src/dtypes.py) mechanism as explained [here](https://jax.readthedocs.io/en/latest/type_promotion.html?highlight=lattice#type-promotion-semantics). The type promotion rules are summarized by the following type lattice:
+
+![JAX type promotion lattice](https://jax.readthedocs.io/en/latest/_images/type_lattice.svg)
+
+
+
+## Dtypes in Linen
+
+Beside input arguments, state and in particular parameters could affect dtype promotion. For example: We might feed a float64 input to a Dense layer with float32 parameters. Currently the result would be truncated to float32. If the input is a complex number the result is even worse because the imaginary part will be silently dropped when casting to float32.
+
+By using the dtype promotion rules already available in JAX we can avoid this issue. A public api is available called `jax.numpy.result_dtype(*args)` which returns the dtype that JAX would promote the given arguments in accordance with the type promotion lattice. For linen layers the arguments would be the layer inputs together with the parameters so for example this would be inputs, kernel, and bias for a linear layer.
+
+
+# Implementation
+
+A simplified example implementation:
+
+
+```python
+def promote_arrays(*xs, dtype):
+ if dtype is None:
+   dtype = jnp.result_type(*jax.tree_leaves(xs))
+ return jax.tree_map(lambda x: jnp.asarray(x, dtype), xs)
+
+Dtype = Any
+class Dense(nn.Module):
+ features: int
+ kernel_init: Callable
+ bias_init: Callable
+ dtype: Optional[Dtype] = None
```

Comment:
I wonder if the current API isn't a bit confusing now. From a user perspective, do you think they will get what the difference is between `param_dtype` and `dtype`? An alternative is to rename `dtype` to `output_dtype`, or is that inaccurate?

Reply:
I would vote for keeping it as proposed, because
1. it's the same naming as before
2. it hints at that the computation is done in `dtype` as opposed to transforming the outputs only

Reply:
That's a big change from a backward's compatibility perspective. I also agree it lacks the computational hint. Anyway, I will add the naming to the discussion section but I will leave it out of the FLIP because we cannot handle too large backward incompatibility in a single change